File: dynamic programming/322. 零钱兑换/java/Coin Change.java
class Solution {
    // 动态规划 背包问题
    public int coinChange(int[] coins, int amount) {
        if (amount == 0) return 0;
        int[] dp = new int[amount + 1];
        for (int i = 1; i <= amount; i++) {
            int min = Integer.MAX_VALUE;
            for (int j = 0; j < coins.length; j++) {
                if (coins[j] <= i) {
                    int val = dp[i - coins[j]] + 1;
                    if (val > 0 && val < min) {
                        min = val;
                    }
                }
            }
            dp[i] = min == Integer.MAX_VALUE ? -1 : min;
        }

        return dp[amount];
    }
}

File: dynamic programming/322. 零钱兑换/go/Coin Change.go
package main

import "sort"

func main() {
	coinChange([]int{1, 2, 5}, 10)
}
// 动态规划 背包问题
func coinChange(coins []int, amount int) int {
	if amount == 0 {
		return 0
	}

	dp := make([]int, amount+1)
	sort.Ints(coins)
	for i := 1; i < len(dp); i++ {
		min := 1 << 32

		for c := 0; c < len(coins) && coins[c] <= i; c++ {
			val := dp[i-coins[c]] + 1
			if val > 0 && val < min {
				min = val
			}
		}

		if min == 1 << 32 {
			dp[i] = -1
		} else {
			dp[i] = min
		}
	}
	return dp[amount]
}
